### DISCUSSION TITLE
'--demangle-cpp' flag, '-Xlinker' interface:  question

### DISCUSSION THREAD

**Discussion by henry2cox** [linux-test-project/lcov#196] [Ideas]
Hi.
I wanted to ask a question of the user community - and also to test whether community discussions will work/how community discussions work.
Please comment if you have an opinion on the feature...or if you don't.

Background:

- some time ago, I had added several callback related options to the lcov/geninfo/genhtml tools.  (See the README and/or the man pages for details.)
- After a while, my (internal) users became a bit more sophisticated - and had callbacks with complicated parameters (e.g., strings with spaces, regexps, etc).
- To relieve 'quoting hell', I modified all of these to behave similar to gcc's -Xlinker option:  pass the option as many times as you like, each appearance is treated as one argument to be passed to the callback.

All good :-)

Then I remembered that the existing "--demangle-cpp" (and related lcovrc options) didn't behave this way :-(

So I went ahead and implemented a local change (not pushed) which aligns the “--demangle” option with the “-Xlinker”-like flag behaviour that I gave to the other (new) flags.

- Basically:  ---demangle takes an optional string argument – and appends that string to an array.
- If no string is provided:  I assume “c++filt” – otherwise, all the strings are joined to produce a command line and command line arguments.
- Empty string (no argument) not allowed except when it is the only arg (i.e., list length ==1) – else error.
- The existing lcovrc ‘genhtml_demangle_cpp_tool’ and ‘genhtml_demangle_ccpp_parad’ are left in place – but marked ‘deprecated’ in the man page
- I also added an $ERROR_DEPRECATED (warning) if it was used.

I'm not sure if anyone would want this new behaviour - or if anyone cares one way or the other.
This could be just making changes for the sake of making changes – and could be simply dropped.
I also think that there are very few users who need demangling _other_ than simple `c++filt`.

Normalizing all the options to act in the same way appeals to my sense of symmetry…but is likely unnecessary, and is overkill.
If done, this WOULD break at least one existing command line:
`$ genhtml -o foo --demangle  myData.info  <- oops…will see ‘myData.info’ as the demangle script, and will complain that there is no tracefile`
Not sure if this is a huge issue or not.

So..what do _you_ think?

Thanks you in advance for your feedback.

Henry

**Comment by henry2cox** [comment]
Just checking whether anyone (apart from me) thinks that this is a Good Idea.
As is:  `--demangle-cpp` is the only callback option which _doesn't_ collect arguments in an -Xlinker-ish way.
I like orthogonality.
OTOH:  it does break backward compatibility in one rather obscure possibly-never-actually-used case.

I'd kind of like to get this into the 2.0 release - which is imminent, I believe.

**Reply by xaizek** [reply]
> I like orthogonality.

Consistency?

> The existing lcovrc ‘genhtml_demangle_cpp_tool’ and ‘genhtml_demangle_ccpp_parad’ are left in place – but marked ‘deprecated’ in the man page

If there is still a way to specify demangler in the config, the change is probably worth it.  2.0 will likely prompt updates and this one will be a  simple one.  If it's gone from the config, the change can be inconvenient to handle though (moving from config to invocation(s)).

**Comment by henry2cox** [comment]
Talked to Peter.  We decided to go ahead with this change - so pushed in c8f83860927e966.
Will be part of the upcoming lcov 2.0 release.
Closing discussion now.
Henry
